perfkitbenchmarker/linux_virtual_machine.py

```diff
@@ -974,3 +975,229 @@ def MoveFile(self, target, source_path, remote_path=''):
+                                       pattern)
+    return output.strip()
+
+  def JujuVersion(self):
+    """Return the Juju version"""
+    output, _ = self.RemoteHostCommand('juju version')
+    return output.strip()
+
+  def JujuSet(self, service, params=[]):
+    """Set the configuration options on a deployed service.
+
+    Args:
+      service: The name of the service
+      params: A list of key=values pairs
+    """
+    output, _ = self.RemoteHostCommand(
+        'juju set %s %s' % (service, ' '.join(params)))
+    return output.strip()
+
+  @vm_util.Retry(poll_interval=30, timeout=3600)
+  def JujuWait(self):
+    """
+    Waits for all deployed services to be installed, configured, and idle.
+
+    Args:
+      poll_interval: How long to wait between checking the current status.
+      timeout: How long to wait for the environment to settle.
+    """
+    status = yaml.load(self.JujuStatus())
+    for service in status['services']:
+      ss = status['services'][service]['service-status']['current']
+
+      # Accept blocked because the service may be waiting on relation
+      if ss not in ['active', 'unknown']:
+          raise errors.Juju.TimeoutException(
+              'Service %s is not ready; status is %s' % (service, ss))
+
+      if ss in ['error']:
+        # The service has failed to deploy.
+        debuglog = self.JujuRun('juju debug-log --limit 200')
+        logging.warn(debuglog)
+        raise errors.Juju.UnitErrorException(
+            'Service %s is in an error state' % service)
+
+      for unit in status['services'][service]['units']:
+        unit_data = status['services'][service]['units'][unit]
+        ag = unit_data['agent-state']
+        if ag != 'started':
+          raise errors.Juju.TimeoutException(
+              'Service %s is not ready; agent-state is %s' % (service, ag))
+
+        ws = unit_data['workload-status']['current']
+        if ws not in ['active', 'unknown']:
+          raise errors.Juju.TimeoutException(
+              'Service %s is not ready; workload-state is %s' % (service, ws))
+
+  def _FindUnit(self, ip):
+    """
+    Find the machine id assigned to a particular IP address
+
+    Args:
+      ip: The IP address of the unit to find
+    """
+    status = yaml.load(self.JujuStatus(ip))
+    if 'machines' in status:
+      return status['machines'].keys()[0]
+    return None
+
+  def JujuDeploy(self, charm, units=1, vm_group=None):
+    """Deploy a charm in a Juju environment
+
+    Args:
+      charm: The charm to deploy, i.e., cs:trusty/ubuntu
+      units: Optional number of units to create.
+      vm_group: The name of vm_group the unit(s) should be deployed to.
+    """
+    machines = []
+    if vm_group:
+      """Deploy (and scale) this service to the machines in its vm group."""
+
+      for unit in self.units:
+        """Find all machines in this vm_group"""
+        if unit.vm_group == vm_group:
+          machines.append(self._FindUnit(unit.internal_ip))
```

Comment:
Not a huge deal, but it might be nice to keep track of machine ids so if you ever need to deploy more than one charm to the same group you don't have to find them again. It would also help if you didn't have to make one juju status call per vm but could just make one call. Up to you if you want to fix it now.